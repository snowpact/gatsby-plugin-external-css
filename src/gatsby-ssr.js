import path from 'path';
import React from 'react';
import fsExtra from 'fs-extra';
import { v1 as uuidv1 } from 'uuid';

const COMPONENT_KEY = 'custom-css';

export const onRenderBody = ({ setHeadComponents }, pluginOptions) => {
  const {
    plugins, // internal to gatsby
    source, // source of the script
    ...options
  } = pluginOptions;

  if (!source) {
    throw new Error(
      'gatsby-plugin-external-css needs a "source" option in order to work correctly',
    );
  }

  const sourceNormalized = path.normalize(source);

  const basename = path.basename(sourceNormalized);

  const fileExists = fsExtra.pathExistsSync(sourceNormalized);
  const isFile = fsExtra.statSync(sourceNormalized).isFile();
  const isCssExtension = path.extname(basename) === '.css';

  if (!fileExists || !isFile || !isCssExtension) {
    throw new Error(
      'gatsby-plugin-external-css needs a "source" option that exists and that is a css file',
    );
  }

  try {
    const newPath = path.join(process.cwd(), 'public', basename);
    fsExtra.copySync(sourceNormalized, newPath);
  } catch (error) {
    throw new Error('gatsby-plugin-external-css could not copy the file into the public folder');
  }

  // Adding custom cache busting
  const pathWithCacheBusting = `${basename}?cb=${uuidv1()}`;
  setHeadComponents([
    <link key={COMPONENT_KEY} rel="stylesheet" type="text/css" href={pathWithCacheBusting} />,
  ]);
};

export const onPreRenderHTML = ({ getHeadComponents, replaceHeadComponents }) => {
  const headComponents = getHeadComponents();

  // Reordering the files so custom css is at the end
  const orderedComponents = headComponents.sort((item) => (item.key === COMPONENT_KEY ? 1 : -1));
  replaceHeadComponents(orderedComponents);
};
