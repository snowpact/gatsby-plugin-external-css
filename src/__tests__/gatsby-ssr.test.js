import { onRenderBody } from '../gatsby-ssr';
import { v1 as uuidv1 } from 'uuid';

jest.mock('uuid');

const setHeadComponents = jest.fn();

const TEST_FILE = `${__dirname}/__modules__/style.css`;
const BAD_TEST_FILE = `${__dirname}/__modules__/bad.scss`;
// const VOID_FILE = `${__dirname}/__modules__/not.css`;

const simpleORB = (pluginOptions) =>
  onRenderBody({ setHeadComponents }, { plugins: [], ...pluginOptions });

beforeAll(() => {
  uuidv1.mockImplementationOnce(() => {
    return 'uuid-123';
  });
});

afterAll(() => {
  uuidv1.mockClear();
});

afterEach(() => {
  setHeadComponents.mockClear();
});

test('pluginsOptions need a "source" property', () => {
  expect(() => simpleORB({})).toThrowErrorMatchingSnapshot('source-attribute');
});

test('pluginsOptions need a valid "source" file css', () => {
  expect(() => simpleORB({ source: BAD_TEST_FILE })).toThrowErrorMatchingSnapshot('source-css');
});

// test('pluginsOptions need a "source" file that exists', () => {
//   expect(() => simpleORB({ source: VOID_FILE })).toThrowErrorMatchingSnapshot('source-exists');
// });

test('Providing default valid source file', () => {
  simpleORB({
    source: TEST_FILE,
  });
  expect(setHeadComponents.mock.calls.length).toBe(1);
  expect(setHeadComponents.mock.calls[0]).toMatchSnapshot('source-valid');
});
