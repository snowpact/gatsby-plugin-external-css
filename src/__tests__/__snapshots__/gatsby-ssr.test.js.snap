// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Providing default valid source file: source-valid 1`] = `
Array [
  Array [
    <link
      href="/style.css?cb=uuid-123"
      rel="stylesheet"
      type="text/css"
    />,
  ],
]
`;

exports[`pluginsOptions need a "source" property: source-attribute 1`] = `"gatsby-plugin-external-css needs a \\"source\\" option in order to work correctly"`;

exports[`pluginsOptions need a valid "source" file css: source-css 1`] = `"gatsby-plugin-external-css needs a \\"source\\" option that exists and that is a css file"`;
